Make volatile opportunistically relaxed

RossSmyth opened the pull request:
This is mainly to test the waters to see if this is something that is even desired.

# What this does

For a single instruction set, make the power of two sizes up to a register width emit single instructions, as one would expect, in asm. I chose thumb because I work on thumbv6m. But it is not possible to choose that specifically. 

For other weird sizes that are basically meaningless in the context of MMIO it is the status quo because there's really nothing more sane that can be done.

I would appreciate if someone could review the asm because while I am an embedded dev, I don't write asm that much.

# The murky past

What is volatile? Who knows! The C standard definitely says something. What is that something is not quite clear. It's all about the _Vibes_. 

> "If something isn't working quite right, chuck a volatile in there"
- Ancient embedded wisdom

Well, what do compilers do? So users aren't unhappy they general do something that matches the vibes. But you can coax them to not in weird edge cases. One that was shown off recently on the Rust Community Discord is this
https://godbolt.org/z/WxP14vqPa

Well the vibes aren't great.

# The controversial present

People like to debate about volatile all the time. What does it do, what do they want it to do, what does it mean for opsem? Well these are all great questions I will not answer. 

Some of the largest users of volatile are embedded developers (disclosure: I am an embedded dev :ferrisclueless:). It should almost exclusively be used for MMIO if the devs are being good. In those cases it is generally desired that the assembly output is a single load. 

Unfortunately libstd make the (imo) mistake that things larger than a register (someone will probably say "erm on X arch you can do MMIO for other things, but that's not the point) can be read with `read_volatile`. But for sizes we can just turn it into a quality of implementation issue. Emit a load. Emit a store. 

# The shiny future

Hopefully people can fill in their favorite arches so Rust's volatile loads and stores are the best they can be. I would also like it if volatile loads and stores were relaxed operations so that people's broken code is slightly less broken. But that's a future discussion.

r? wg-embedded-arm

RossSmyth committed:
Move read_volatile to volatile module and implement thumbv6 in asm.

The goal is to ensure LLVM doesn't do anything
silly (:3) with the loads.

RossSmyth committed:
Move write_volatile to the volatile module

RossSmyth committed:
Add thumb asm volatile writes

RossSmyth commented:
r? arm

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14234624018/job/39891621790) [(plain)](https://github.com/rust-lang/rust/commit/220ad1161f04c814ae4e2b4895fa6e47c46e8384/checks/39891621790/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
  make \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 3.342 Building wheels for collected packages: reuse
#12 3.343   Building wheel for reuse (pyproject.toml): started
#12 3.553   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.554   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=5bb60f62728aaedff7162745ce743c7f2f55069b3e7f82e6a37d70df455797cc
#12 3.555   Stored in directory: /tmp/pip-ephem-wheel-cache-8h5em9dc/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.557 Successfully built reuse
#12 3.557 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.949 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.949 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.498 Collecting virtualenv
#12 4.565   Downloading virtualenv-20.30.0-py3-none-any.whl (4.3 MB)
#12 4.763      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 22.1 MB/s eta 0:00:00
#12 4.826 Collecting platformdirs<5,>=3.9.1
#12 4.839   Downloading platformdirs-4.3.7-py3-none-any.whl (18 kB)
#12 4.886 Collecting filelock<4,>=3.12.2
#12 4.897   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.921 Collecting distlib<1,>=0.3.7
#12 4.931   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.947      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 33.9 MB/s eta 0:00:00
#12 5.029 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 5.212 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.7 virtualenv-20.30.0
#12 5.212 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 5.3s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      124864 kB
DirectMap2M:     8263680 kB
DirectMap1G:    10485760 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
[TIMING] core::build_steps::tool::LibcxxVersionTool { target: x86_64-unknown-linux-gnu } -- 0.216
---
fmt check
fmt: checked 5939 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/library/core/src/ptr/write_volatile.md:38: undocumented unsafe
##[error]tidy error: /checkout/library/core/src/ptr/read_volatile.md:39: undocumented unsafe
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.0.1)
linting python files
All checks passed!
checking python file formatting
26 files already formatted
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:48
  local time: Thu Apr  3 04:09:51 UTC 2025
  network time: Thu, 03 Apr 2025 04:09:51 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

KittyBorgX commented:
r? arm

Noratrieb commented:
@rust-lang/opsem This came out of a discussion about volatile not being usable for MMIO in some cases, where the (virtualized) implementation does not support MMIO through all instructions, but only some basic ones. What do you think about this?

RalfJung commented:
Sorry, could someone explain the context for non-embedded people here?

> Well, what do compilers do? So users aren't unhappy they general do something that matches the vibes. But you can coax them to not in weird edge cases. One that was shown off recently on the Rust Community Discord is this
https://godbolt.org/z/WxP14vqPa

This example looks like the output of pwgen to my eyes. What is it supposed to tell me? :)

RalfJung commented:
FWIW there have been many hundreds of comments in various issues around "volatile" on the UCG and opsem issue tracker and Zulip. We'd be happy to work with someone who wants to put in the time of properly re-doing our entire volatile story, but unfortunately we don't have the resources to do it entirely by ourselves.

"It's basically inline assembly" is IMO the most promising approach, though it is unfortunate that apparently we cannot rely on LLVM doing that.

RalfJung commented on library/core/src/ptr/read_volatile.md:
I don't understand the point of these macro contraptions and why the docs are now in a separate file. Can't we  have a single function with a `cfg_match!` inside? Why does the function itself need to be emitted by a macro?

CAD97 replied:
Probably what resulted in this is that naive use of `cfg_match!`'s current implementation doesn't behave great in an expression position that wants to produce a value. That, or being first written with a `cfg_if!` that requires the arms to be made up of items.

You can write `cfg_match! {{ … }}` with two brackets to get it to consistently work as an expression block.

RossSmyth replied:
I just found this way more readable. And macros and doc comments don't interact well.

RalfJung replied:
It's quite hard to read to have the comment out-of-line for *one* of many functions. And the macro shouldn't have anything to do with the doc comment in the first place. The macro should generate the *body* of `write_volatile`, not the entire function.

RalfJung commented on library/core/src/ptr/volatile.rs:
```diff
@@ -0,0 +1,134 @@
+                    dest = in(reg) dst,
+                ),
+                4 => asm!(
+                    "str {val}, [{dest}]",
```
Is it meaningful that `str` etc are lower-case but `LDR` etc are upper-case?

Noratrieb commented:
@RalfJung the way this came up in discussion again on the community discord is that someone was having problems with their non-embedded x86-64 OS in KVM (virtualization). KVM apparently only allows MMIO to happen with a specific subset of instructions, and LLVM sometimes codegened the volatile with other instructions (merging it with operations for example) such that while the write was still present, KVM no longer understood it.
What they'd have needed is for volatile to not just guarantee that the write isn't elided, and not just that it doesn't tear, but also that it compiles to a "normal, simple" (I'd guess a mov in this case for x86-64) instruction that KVM can understand.

RalfJung commented:
The PR does not change anything for x86 though?

I think ideally this would be fixed in LLVM or KVM. The two do not agree on how to do MMIO. We can work around that on the Rust side, but not without at least trying to fix the actual issue. Has this problem been reported with either of them?

saethlin commented:
> What do you think about this, guaranteeing that volatile accesses turn into specific instructions for specific sized?

I do not think we should guarantee anything. In the referenced discussion I encouraged a PR such as this (though I'll admit I didn't anticipate so much diff) as a pure quality-of-implementation concern. No changes to any documentation.

But in addition, since people like to read the standard library and take its internal comments as normative, we will need to be very clear in the implementation here to "document" with internal comments that any particular codegen is not guaranteed.

RossSmyth commented:
> Why is it not possible to choose...? You seem to be doing quite well choosing it for this PR. confused
It uses cfg to select arm, thumb, 32-bit. Which is a wider range that thumv6m.

> This example looks like the output of pwgen to my eyes. What is it supposed to tell me? :)
Many people expect volatile not to merge anything. This does. 

> The PR does not change anything for x86 though?
Yes, this is a first step to see if people even want to see a change like this. Rather than touching the precious x86, this just touches an embedded platform. 

> I think ideally this would be fixed in LLVM
Probably! But I will not be the one to do that. 

From my POV LLVM is great at optimizing code, but it seems that all the different passes often forget things. So saying "well fix it in LLVM," means that LLVM should not forget things throughout its pipeline. Which would be great! But history has suggested that is a hard problem.

RalfJung commented:
> Many people expect volatile not to merge anything. This does.

What is being merged? With what? Where? Again, your example is entirely unreadable to me. So please explain whatever you think the example should demonstrate.

> Probably! But I will not be the one to do that.

I don't think we should merge anything here without at least filing an issue on the LLVM side. And if LLVM considers this not-a-bug, we'll have to determine if we are ready to carry our own implementation of this ~forever.

RalfJung commented:
> Another funny idea is to internally tell llvm it is an atomic relaxed volatile load 

Yes, I want all our volatile accesses to compile to atomic accesses in LLVM. However, sadly we allow volatile accesses on all types, so we'll have to figure something out for too-large and misaligned accesses.

So if the issue you are having can be fixed by marking `volatile` accesses as atomic for types that are sufficiently small and where size==align, that would be a much better fix than the current proposal IMO.

RossSmyth committed:
Fix links

CAD97 commented on library/core/src/ptr/write_volatile.md:
This file is now redundant again and should be removed.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14248003554/job/39933791475) [(plain)](https://github.com/rust-lang/rust/commit/2de7d139512c3810933e266a95592a8560fee9ef/checks/39933791475/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
  make \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 3.081 Building wheels for collected packages: reuse
#12 3.083   Building wheel for reuse (pyproject.toml): started
#12 3.294   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.295   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=5bb60f62728aaedff7162745ce743c7f2f55069b3e7f82e6a37d70df455797cc
#12 3.295   Stored in directory: /tmp/pip-ephem-wheel-cache-1p0c6nhv/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.297 Successfully built reuse
#12 3.298 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.690 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.690 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.254 Collecting virtualenv
#12 4.326   Downloading virtualenv-20.30.0-py3-none-any.whl (4.3 MB)
#12 4.413      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 51.7 MB/s eta 0:00:00
#12 4.480 Collecting platformdirs<5,>=3.9.1
#12 4.492   Downloading platformdirs-4.3.7-py3-none-any.whl (18 kB)
#12 4.517 Collecting distlib<1,>=0.3.7
#12 4.529   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.536      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 95.3 MB/s eta 0:00:00
#12 4.579 Collecting filelock<4,>=3.12.2
#12 4.590   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.671 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.853 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.7 virtualenv-20.30.0
#12 4.853 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 4.9s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      114624 kB
DirectMap2M:     7225344 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
[TIMING] core::build_steps::tool::LibcxxVersionTool { target: x86_64-unknown-linux-gnu } -- 0.218
---
fmt check
fmt: checked 5939 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/library/core/src/ptr/write_volatile.md:38: undocumented unsafe
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.0.1)
linting python files
All checks passed!
checking python file formatting
26 files already formatted
checking C++ file formatting
some tidy checks failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:49
  local time: Thu Apr  3 16:36:41 UTC 2025
  network time: Thu, 03 Apr 2025 16:36:41 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

CAD97 commented on library/core/src/ptr/volatile.rs:
```diff
@@ -0,0 +1,240 @@
+                is_zst: bool = T::IS_ZST,
+            ) => crate::ub_checks::maybe_is_aligned_and_not_null(addr, align, is_zst)
+        );
+        cfg_match! {
```
Of course ideally all of this magic would happen in the intrinsic / in LLVM. *If* we're going to implement volatile semantics in lib code for any appreciable amount of time, though, it should probably be done in sys PAL modules rather than inline due to the quantity of different platforms which will likely want to provide `asm!` volatile semantics.

RalfJung replied:
`sys` is a std thing, we can't use it here. And anyway it's for abstracting OS/platform differences, not CPU architecture differences.

CAD97 reviewed (commented):
With my opsem hat, I'm with Ralf here: making Rust's "volatile" mean "volatile atomic" for "appropriately sized" accesses better matches what developers generally expect when they reach for `volatile`. But also that if LLVM isn't preserving the exact read size of "appropriately sized" volatile reads, that this is an LLVM bug and we should attempt to fix it in LLVM before working around it, or at least link to and track an LLVM issue tracking a fix so that we can remove the workaround once it isn't needed anymore.

saethlin commented:
The only libs reviewer here is @Noratrieb. There is no libs reviewer assigned, and libs hasn't been contacted generally, only opsem. I very strongly prefer that questions about the feasibility of carrying a library implementation such as this (as well as guidance on how to implement this well) come from a libs maintainer.

@rust-lang/libs

RossSmyth committed:
asm -> atomic_relaxed
TODO: atomic_volatile_relaxed

RossSmyth committed:
Use atomic volatile relaxed intrinsic in the implementation of volatile.

RossSmyth committed:
Add new volatile relaxed atomic intrinsic

Copied the code from here:
https://github.com/llvm/llvm-project/blob/24dfcc0c024f9ab8ba61c0994513f57e882961fc/llvm/lib/Frontend/Atomic/Atomic.cpp#L31-L32

Did not implement it for GCC or craneline.

RossSmyth commented:
Ok here it is with volatile atomics.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/14274454689/job/40014747894) [(plain)](https://github.com/rust-lang/rust/commit/2628b10be76b51f6f52e0b50a7306955b87dc537/checks/40014747894/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
info: removing rustup binaries
info: rustup is uninstalled
##[group]Image checksum input
mingw-check-tidy
# We use the ghcr base image because ghcr doesn't have a rate limit
# and the mingw-check-tidy job doesn't cache docker images in CI.
FROM ghcr.io/rust-lang/ubuntu:22.04

ARG DEBIAN_FRONTEND=noninteractive
RUN apt-get update && apt-get install -y --no-install-recommends \
  g++ \
  make \
---

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
#
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#12 2.847 Building wheels for collected packages: reuse
#12 2.849   Building wheel for reuse (pyproject.toml): started
#12 3.072   Building wheel for reuse (pyproject.toml): finished with status 'done'
#12 3.073   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132719 sha256=5bb60f62728aaedff7162745ce743c7f2f55069b3e7f82e6a37d70df455797cc
#12 3.073   Stored in directory: /tmp/pip-ephem-wheel-cache-nbokbbr7/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#12 3.076 Successfully built reuse
#12 3.076 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#12 3.496 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#12 3.496 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 4.059 Collecting virtualenv
#12 4.097   Downloading virtualenv-20.30.0-py3-none-any.whl (4.3 MB)
#12 4.190      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 4.3/4.3 MB 48.1 MB/s eta 0:00:00
#12 4.230 Collecting distlib<1,>=0.3.7
#12 4.234   Downloading distlib-0.3.9-py2.py3-none-any.whl (468 kB)
#12 4.241      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 469.0/469.0 KB 91.6 MB/s eta 0:00:00
#12 4.275 Collecting platformdirs<5,>=3.9.1
#12 4.279   Downloading platformdirs-4.3.7-py3-none-any.whl (18 kB)
#12 4.315 Collecting filelock<4,>=3.12.2
#12 4.320   Downloading filelock-3.18.0-py3-none-any.whl (16 kB)
#12 4.404 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#12 4.606 Successfully installed distlib-0.3.9 filelock-3.18.0 platformdirs-4.3.7 virtualenv-20.30.0
#12 4.607 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#12 DONE 4.7s

#13 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#13 DONE 0.0s
---
DirectMap4k:      118720 kB
DirectMap2M:     5124096 kB
DirectMap1G:    13631488 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py,cpp
##[group]Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.05s
##[endgroup]
WARN: currently no CI rustc builds have rustc debug assertions enabled. Please either set `rust.debug-assertions` to `false` if you want to use download CI rustc or set `rust.download-rustc` to `false`.
[TIMING] core::build_steps::tool::LibcxxVersionTool { target: x86_64-unknown-linux-gnu } -- 0.231
---
fmt check
fmt: checked 5942 files
tidy check
tidy: Skipping binary file check, read-only filesystem
##[error]tidy error: /checkout/library/core/src/ptr/volatile.rs:84: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
##[error]tidy error: /checkout/library/core/src/ptr/volatile.rs:191: TODO is used for tasks that should be done before merging a PR; If you want to leave a message in the codebase use FIXME
removing old virtual environment
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'venv'
creating virtual environment at '/checkout/obj/build/venv' using 'python3.10' and 'virtualenv'
Requirement already satisfied: pip in ./build/venv/lib/python3.10/site-packages (25.0.1)
linting python files
All checks passed!
checking python file formatting
26 files already formatted
checking C++ file formatting
/checkout/compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp:607:68: error: code should be clang-formatted [-Wclang-format-violations]
                        const char *Name, LLVMAtomicOrdering Order, LLVMBool isVolatile) {
                                                                   ^
/checkout/compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp:618:24: error: code should be clang-formatted [-Wclang-format-violations]
extern "C" LLVMValueRef LLVMRustBuildAtomicStore(LLVMBuilderRef B,
                       ^
/checkout/compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp:618:67: error: code should be clang-formatted [-Wclang-format-violations]
extern "C" LLVMValueRef LLVMRustBuildAtomicStore(LLVMBuilderRef B,
                                                                  ^
/checkout/compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp:619:65: error: code should be clang-formatted [-Wclang-format-violations]
                                                 LLVMValueRef V,
                                                                ^
/checkout/compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp:620:70: error: code should be clang-formatted [-Wclang-format-violations]
                                                 LLVMValueRef Target,
                                                                     ^
/checkout/compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp:621:75: error: code should be clang-formatted [-Wclang-format-violations]
                                                 LLVMAtomicOrdering Order,
                                                                          ^

clang-format linting failed! Printing diff suggestions:
--- /checkout/compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp (actual)
+++ /checkout/compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp (formatted)
@@ -603,9 +603,10 @@
 }
 
 extern "C" LLVMValueRef
 LLVMRustBuildAtomicLoad(LLVMBuilderRef B, LLVMTypeRef Ty, LLVMValueRef Source,
-                        const char *Name, LLVMAtomicOrdering Order, LLVMBool isVolatile) {
+                        const char *Name, LLVMAtomicOrdering Order,
+                        LLVMBool isVolatile) {
   Value *Ptr = unwrap(Source);
   LoadInst *LI = unwrap(B)->CreateLoad(unwrap(Ty), Ptr, Name);
   LI->setAtomic(fromRust(Order));
 
@@ -614,13 +615,11 @@
     LI->setVolatile(true);
   return wrap(LI);
 }
 
-extern "C" LLVMValueRef LLVMRustBuildAtomicStore(LLVMBuilderRef B,
-                                                 LLVMValueRef V,
-                                                 LLVMValueRef Target,
-                                                 LLVMAtomicOrdering Order,
-                                                 LLVMBool isVolatile) {
+extern "C" LLVMValueRef
+LLVMRustBuildAtomicStore(LLVMBuilderRef B, LLVMValueRef V, LLVMValueRef Target,
+                         LLVMAtomicOrdering Order, LLVMBool isVolatile) {
   StoreInst *SI = unwrap(B)->CreateStore(unwrap(V), unwrap(Target));
   SI->setAtomic(fromRust(Order));
 
   // atomic volatile

some tidy checks failed
tidy error: checks with external tool 'clang-format' failed
Command has failed. Rerun with -v to see more details.
Build completed unsuccessfully in 0:01:51
  local time: Fri Apr  4 21:37:05 UTC 2025
  network time: Fri, 04 Apr 2025 21:37:05 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

Amanieu commented:
I don't believe this change belong in the standard library. It would be very surprising for `volatile` in Rust to work differently than `volatile` in Clang, so if this change has merit then it should be implemented directly in LLVM so that other languages can benefit.

With that said I did review the Discord discussion and I don't think changing the `volatile` codegen is the correct solution here. Fundamentally the issue is that the KVM hypervisor handles virtual MMIO by causing instructions accessing an MMIO region to trap and then decoding the trapping instruction to figure out what it does. This means that to work with KVM, the instruction which accesses MMIO must be one of the memory access instructions that KVM can decode. This is a much strong guarantee than what `volatile` provides, even in C, and you really should be using inline assembly in your code instead of volatile operations.

I think this is a good opportunity for a MMIO crate which explicitly guarantees that it emits KVM-compatible instructions on supported architectures.

RossSmyth commented:
I think opportunistically making volatile to be relaxed atomic when it is able to would be better. We are allowed to be better than Clang. And I doubt LLVM would accept making volatile atomic. 

It would be much clearer to the many people who ask what volatile means wrt atomics, and people do and will continue to write code assuming volatile means something in terms of synchronization, so it would mean their code is less broken.

Amanieu commented:
I don't think this actually makes things better, it just encourages people to rely on stronger guarantees than what we actually provide. The definition of `volatile` in Rust has always been "whatever C does" so it doesn't make sense to diverge from C here.

taiki-e commented:
FWIW, my [atomic-maybe-uninit](https://github.com/taiki-e/atomic-maybe-uninit) crate uses "volatile" inline assembly and provides at least atomic load/store on almost all CPU architectures supported by Rust.

It was originally created for [use with crossbeam](https://github.com/crossbeam-rs/crossbeam/pull/1015), but I guess it can be used for this purpose as well, since relaxed atomic load stores up to at least a pointer width use normal load/store instructions.

The functionality this crate provides is something that is [considered to be impossible to implement without inline assembly at this time](https://github.com/rust-lang/unsafe-code-guidelines/issues/449), and since there are so few instruction choices reasonably available for within-register-size integer relaxed atomic load/store, there should be no problem in providing more specific implementation guarantees, including instructions that are actually used.

RalfJung commented:
> It would be very surprising for volatile in Rust to work differently than volatile in Clang

`volatile` in clang (and C more broadly) is notoriously under-defined. I don't think that should stop us from improving the situation in Rust.

Indeed ideally this should be done by fixing our definition from "whatever C does" to something actually well-defined.
